include/curl/curl.h

```diff
@@ -748,6 +748,31 @@ typedef CURLcode (*curl_ssl_ctx_callback)(CURL *curl,    /* easy handle */
                                                           mbedtls_ssl_config */
                                           void *userptr);
 
+struct ssl_session_dump {
+  int backend;             /* curl_sslbackend enum value */
+  const char *scheme;      /* protocol scheme used */
+  const char *hostname;    /* host name for which this ID was used */
+  int port;                /* remote port */
+  unsigned char *sessionblob; /* serialized session data */
+  size_t blobsize;         /* session data size */
+};
+
+typedef CURLcode (*curl_ssl_cache_callback)(CURL *curl,
+                                          /* user code must allocate all mem
+                                             for *dump and its members, while
+                                             curl will release it */
+                                          struct ssl_session_dump **dump,
+                                          size_t *size,
+                                          /* set the release call for memory
+                                             allocated in user code, and
+                                             guaranteed to apply to the same
+                                             heap (typically *free_cb = free,
+                                             when malloc() was used); if not
+                                             set, Curl_cfree() will be used
+                                             instead. */
+                                          curl_free_callback *free_cb,
+                                          void *userptr);
```

Comment:
This is *really* complicated and is just asking for users to mess things up. We need a simpler way!

Reply:
Yes, I totally agree. The problem was that I could not come up with a simpler way. One alternative would be to drop `free_cb` argument and require the use of `curl_global_init_mem()` for this. But that did not sound like a lot less complicated and is probably even less obvious. Should I change it like that?

Otherwise, is there a way to request the lib for an allocated structure [collection] that it would then also dealocate? something like `void* curl_alloc_type(SOME_ENUM_TYPE_ID id, size_t count)` ?
I'd say this is one the most important problems that needs to be settled for this patch.